src/rust/cpu/cpu.rs

```diff
@@ -1967,6 +1997,47 @@ pub unsafe fn do_page_walk(
     return Ok(high);
 }
 
+unsafe fn walk_page_directory(pae: bool, addr: i32) -> Option<(i32, i32)> {
+    if pae {
+        let pdpt_idx = (addr as u32) >> 30;
+        let page_dir_idx = ((addr as u32) >> 21) & 0x1FF;
+
+        let pdpt_addr = (*cr.offset(3) as u32 >> 2).wrapping_add(pdpt_idx << 1);
+        let pdpt_entry = read_aligned32(pdpt_addr);
+        if pdpt_entry & PAGE_TABLE_PRESENT_MASK == 0 {
+            return None;
+        }
+
+        let page_dir_addr = ((pdpt_entry as u32 & 0xFFFFF000)>> 2).wrapping_add(page_dir_idx << 1);
+        let page_dir_entry = read_aligned32(page_dir_addr);
+        return Some((page_dir_addr as i32, page_dir_entry));
+    }
+
+    let page_dir_idx = (addr as u32) >> 22;
+    let page_dir_addr = (*cr.offset(3) as u32 >> 2).wrapping_add(page_dir_idx);
+    let page_dir_entry = read_aligned32(page_dir_addr);
+    return Some((page_dir_addr as i32, page_dir_entry));
+}
+
+unsafe fn walk_page_table(
+    pae: bool,
+    addr: i32,
+    page_dir_entry: i32
+) -> (i32, i32) {
+    let page_table = (page_dir_entry as u32 & 0xFFFFF000) >> 2;
+    if pae {
+        let page_table_idx = (addr as u32 >> 12) & 0x1FF;
+        let page_table_addr = page_table.wrapping_add(page_table_idx << 1);
+        let page_table_entry = read_aligned32(page_table_addr);
+        return (page_table_addr as i32, page_table_entry);
```

Comment:
This is where the 64-bit address is read from, is that correct?

Could you change it to something like:

```
        let page_table_entry = read_aligned64(page_table_addr);
        dbg_assert!(page_table_entry & 0xFFFF_FFFF_0000_0000 == 0, "Unsupported: Physical 64-bit address");
        return (page_table_addr as i32, page_table_entry as i32);
```

Reply:
Done.

There's no `read_aligned64` and I wasn't sure what the convention for the argument should be. I went with "byte address >> 2", for consistency with `read_aligned32` (using "byte address >> 3" sounded like a more confusing option).

The highest bit of PDE and PTE (but not the highest entry, PDPTE) is execute-disable and can be set. I added comments about it.